common/test/com/commercetools/sunrise/framework/viewmodels/content/products/SelectableProductAttributeViewModelFactoryTest.java

```diff
@@ -0,0 +1,122 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Unit tests for {@link SelectableProductAttributeViewModelFactory}.
+ */
+public class SelectableProductAttributeViewModelFactoryTest extends WithApplication {
+    private static final String STRING_ATTRIBUTE_NAME = "stringAttribute";
+    private static final String NUMBER_ATTRIBUTE_NAME = "numberAttribute";
+    private static final String PRODUCT_TYPE_ID = "productTypeId";
+
+    private ProductAttributeSettings fakeProductAttributeSettings;
+    private ProductTypeLocalRepository fakeProductTypeLocalRepository;
+
+    private SelectableProductAttributeViewModelFactory viewModelFactory;
+
+    private Attribute stringAttribute1;
+    private Attribute stringAttribute2;
+    private Attribute numberAttribute1;
+    private Attribute numberAttribute2;
+
+    private ProductType fakeProductType;
+    private Reference<ProductType> productTypeRef;
+
+    @Before
+    public void setup() {
+        viewModelFactory = app.injector().instanceOf(SelectableProductAttributeViewModelFactory.class);
+        fakeProductType = mock(ProductType.class);
+
+        final AttributeDefinition stringAttributeDefinition =
+                AttributeDefinitionBuilder.of(STRING_ATTRIBUTE_NAME, LocalizedString.of(Locale.ENGLISH, "string"), StringAttributeType.of())
+                        .build();
+        when(fakeProductType.getAttribute(STRING_ATTRIBUTE_NAME)).thenReturn(stringAttributeDefinition);
+        when(fakeProductType.findAttribute(STRING_ATTRIBUTE_NAME)).thenReturn(Optional.of(stringAttributeDefinition));
+        stringAttribute1 = Attribute.of(STRING_ATTRIBUTE_NAME, AttributeAccess.ofString(), "v1");
+        stringAttribute2 = Attribute.of(STRING_ATTRIBUTE_NAME, AttributeAccess.ofString(), "v2");
+
+        final AttributeDefinition numberAttributeDefinition =
+                AttributeDefinitionBuilder.of(NUMBER_ATTRIBUTE_NAME, LocalizedString.of(Locale.ENGLISH, "number"), NumberAttributeType.of())
+                        .build();
+        when(fakeProductType.getAttribute(NUMBER_ATTRIBUTE_NAME)).thenReturn(numberAttributeDefinition);
+        when(fakeProductType.findAttribute(NUMBER_ATTRIBUTE_NAME)).thenReturn(Optional.of(numberAttributeDefinition));
+
+        numberAttribute1 = Attribute.of(NUMBER_ATTRIBUTE_NAME, AttributeAccess.ofDouble(), 12.0);
+        numberAttribute2 = Attribute.of(NUMBER_ATTRIBUTE_NAME, AttributeAccess.ofDouble(), 13.0);
+
+        when(fakeProductType.getId()).thenReturn(PRODUCT_TYPE_ID);
+        when(fakeProductType.toReference()).thenReturn(productTypeRef);
+        when(fakeProductType.getAttributes()).thenReturn(Arrays.asList(stringAttributeDefinition, numberAttributeDefinition));
+        productTypeRef = Reference.ofResourceTypeIdAndObj(ProductType.resourceTypeId(), fakeProductType);
+        when(fakeProductTypeLocalRepository.findById(PRODUCT_TYPE_ID)).thenReturn(Optional.of(fakeProductType));
```

Comment:
This looks like a lot of work for just one test. My biggest fear is that it might not be easily scalable for more tests, and we are in fact missing many tests cases.  In the end I think we should have one test where it is clear what is encoded and what is not (with a single attribute) and then try with combination of attributes and edge cases (many attributes-single value, single attribute-multiple value).
Also another suggestion would be to use JSON representations in separate files of a real responses from the platform, and having them in each test.

Reply:
I agree with you, let's see if we can find a better solution together!